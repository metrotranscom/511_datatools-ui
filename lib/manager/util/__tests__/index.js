// @flow

import clone from 'lodash/cloneDeep'
import moment from 'moment'

import {feedSortOptions, parseBounds, versionStatusFilters} from '../'
import {
  restoreDateNowBehavior,
  setTestTime
} from '../../../../__tests__/test-utils'
import mockData from '../../../../__tests__/test-utils/mock-data'

import type {Feed} from '../../../types'

const {mockFeedWithVersion, mockFeedWithoutVersion} = mockData.manager

// create a third mock feed for sorting
const mockFeedWithVersion2 = clone(mockFeedWithVersion)
mockFeedWithVersion2.id = 'another test feed with a version that ends later'
mockFeedWithVersion2.name = 'another test feed with a version that ends later'
mockFeedWithVersion2.lastUpdated += 86400000
// it exists, but flow needs to know for sure
if (mockFeedWithVersion2.latestValidation) {
  mockFeedWithVersion2.latestValidation.errorCount = 12345
  mockFeedWithVersion2.latestValidation.startDate = '20190801'
  mockFeedWithVersion2.latestValidation.endDate = '20200801'
}

const mockFeeds = [
  mockFeedWithVersion,
  mockFeedWithVersion2,
  mockFeedWithoutVersion
]

function getFeedName (feed) {
  return feed.name
}

function setFeedValidDateName (feed: Feed): Feed {
  const newFeed = clone(feed)
  if (newFeed.latestValidation) {
    newFeed.name = `test feed valid from ${newFeed.latestValidation.startDate} - ${newFeed.latestValidation.endDate}`
  } else {
    newFeed.name = 'test feed with no version'
  }
  return newFeed
}

describe('manager > util > index >', () => {
  describe('parseBounds >', () => {
    describe('invalid >', () => {
      it('should fail to parse less than 3 values', () => {
        expect(parseBounds('1,2,3')).toMatchSnapshot()
      })

      it('should fail to parse invalid lat/lon coordinates', () => {
        expect(parseBounds('-200,-100,210,120')).toMatchSnapshot()
      })
    })

    describe('valid >', () => {
      it('should parse valid bounds', () => {
        expect(parseBounds('-122.723842, 45.503841, -122.696827, 45.525433')).toMatchSnapshot()
      })
    })
  })

  describe('versionStatusFilters', () => {
    const mockFeedsForSortTest = mockFeeds.map(setFeedValidDateName)

    afterEach(restoreDateNowBehavior)

    const filterTestCaseData = {
      'all': {
        expected: [
          'test feed valid from 20180801 - 20190801',
          'test feed valid from 20190801 - 20200801',
          'test feed with no version'
        ]
      },
      'active': {
        expected: ['test feed valid from 20180801 - 20190801'],
        mockDate: Date.UTC(2019, 0, 1)
      },
      'expiring': {
        expected: ['test feed valid from 20180801 - 20190801'],
        mockDate: Date.UTC(2019, 6, 19)
      },
      'expired': {
        expected: ['test feed valid from 20180801 - 20190801'],
        mockDate: Date.UTC(2020, 6, 19)
      },
      'future': {
        expected: ['test feed valid from 20190801 - 20200801'],
        mockDate: Date.UTC(2019, 6, 30)
      }
    }

    Object.keys(versionStatusFilters).forEach(optionKey => {
      const testCaseData = filterTestCaseData[optionKey]
      const filterOptionTestDate = testCaseData.mockDate
        ? testCaseData.mockDate
        : Date.UTC(2018, 6, 30)

      /**
       * Make sure filtering happens as expected. The snapshot created is based
       * off of an array of just the feed names.
       */
      it(`should filter feeds with '${optionKey}' strategy for date ${moment(filterOptionTestDate).toISOString()}`, () => {
        // override Date library to set a test time before each type of test
        setTestTime(filterOptionTestDate)

        const filtered = clone(mockFeedsForSortTest)
          .filter(feed => versionStatusFilters[optionKey](feed.latestValidation))
          .map(getFeedName)
        expect(filtered).toHaveLength(testCaseData.expected.length)
        expect(filtered).toEqual(expect.arrayContaining(testCaseData.expected))
      })
    })
  })

  describe('feedSortOptions', () => {
    Object.keys(feedSortOptions).forEach(optionKey => {
      /**
       * Make sure sorting happens as expected. The snapshot created is based
       * off of an array of just the feed names.
       */
      it(`should sort ${optionKey}`, () => {
        expect(
          clone(mockFeeds)
            .sort(feedSortOptions[optionKey])
            .map(getFeedName)
        ).toMatchSnapshot()
      })
    })
  })
})
