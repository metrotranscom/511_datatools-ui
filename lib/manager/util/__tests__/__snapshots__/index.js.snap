// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > util > index > feedSortOptions should sort alphabetically-asc 1`] = `
Array [
  "another test feed with a version that ends later",
  "test feed with a version",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort alphabetically-desc 1`] = `
Array [
  "test feed with no version",
  "test feed with a version",
  "another test feed with a version that ends later",
]
`;

exports[`manager > util > index > feedSortOptions should sort endDate-asc 1`] = `
Array [
  "test feed with a version",
  "another test feed with a version that ends later",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort endDate-desc 1`] = `
Array [
  "another test feed with a version that ends later",
  "test feed with a version",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort lastUpdated-asc 1`] = `
Array [
  "test feed with a version",
  "another test feed with a version that ends later",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort lastUpdated-desc 1`] = `
Array [
  "another test feed with a version that ends later",
  "test feed with a version",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort numErrors-asc 1`] = `
Array [
  "test feed with a version",
  "another test feed with a version that ends later",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort numErrors-desc 1`] = `
Array [
  "another test feed with a version that ends later",
  "test feed with a version",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort startDate-asc 1`] = `
Array [
  "test feed with a version",
  "another test feed with a version that ends later",
  "test feed with no version",
]
`;

exports[`manager > util > index > feedSortOptions should sort startDate-desc 1`] = `
Array [
  "another test feed with a version that ends later",
  "test feed with a version",
  "test feed with no version",
]
`;

exports[`manager > util > index > parseBounds > invalid > should fail to parse invalid lat/lon coordinates 1`] = `
Object {
  "errors": Array [
    "West value is invalid. Expected a value between 180 and -180, but found -200",
    "East value is invalid. Expected a value between 180 and -180, but found 210",
    "North value is invalid. Expected a value between 90 and -90, but found 120",
    "South value is invalid. Expected a value between 90 and -90, but found -100",
  ],
  "valid": false,
}
`;

exports[`manager > util > index > parseBounds > invalid > should fail to parse less than 3 values 1`] = `
Object {
  "errors": Array [
    "Incorrect amount of values. Found 3, expected 4.",
  ],
  "valid": false,
}
`;

exports[`manager > util > index > parseBounds > valid > should parse valid bounds 1`] = `
Object {
  "bounds": Object {
    "east": -122.696827,
    "north": 45.525433,
    "south": 45.503841,
    "west": -122.723842,
  },
  "valid": true,
}
`;
